tests/core/contracts/test_contract_constructor.py

```diff
@@ -0,0 +1,194 @@
+        constructor_kwargs):
+    gas_estimate = WithConstructorArgumentsContract.constructor(
+        *constructor_args, **constructor_kwargs).estimateGas()
+    try:
+        assert abs(gas_estimate - 78572) < 200  # Geth
+    except AssertionError:
+        assert abs(gas_estimate - 112574) < 200  # eth-tester with py-evm
+
+
+def test_contract_constructor_gas_estimate_with_constructor_with_address_argument(
+        WithConstructorAddressArgumentsContract):
+    gas_estimate = WithConstructorAddressArgumentsContract.constructor(
+        "0x16D9983245De15E7A9A73bC586E01FF6E08dE737").estimateGas()
+    try:
+        assert abs(gas_estimate - 50181) < 200  # Geth
+    except AssertionError:
+        assert abs(gas_estimate - 95399) < 200  # eth-tester with py-evm
+
+
+def test_contract_constructor_transact_no_constructor(web3, MathContract, MATH_RUNTIME):
+    deploy_txn = MathContract.constructor().transact()
+
+    txn_receipt = web3.eth.getTransactionReceipt(deploy_txn)
+    assert txn_receipt is not None
+
+    assert txn_receipt['contractAddress']
+    contract_address = txn_receipt['contractAddress']
+
+    blockchain_code = web3.eth.getCode(contract_address)
+    assert blockchain_code == decode_hex(MATH_RUNTIME)
+
+
+def test_contract_constructor_transact_with_constructor_without_arguments(
+        web3, SimpleConstructorContract, SIMPLE_CONSTRUCTOR_RUNTIME):
+    deploy_txn = SimpleConstructorContract.constructor().transact()
+
+    txn_receipt = web3.eth.getTransactionReceipt(deploy_txn)
+    assert txn_receipt is not None
+
+    assert txn_receipt['contractAddress']
+    contract_address = txn_receipt['contractAddress']
+
+    blockchain_code = web3.eth.getCode(contract_address)
+    assert blockchain_code == decode_hex(SIMPLE_CONSTRUCTOR_RUNTIME)
+
+
+@pytest.mark.parametrize(
+    'constructor_args,constructor_kwargs',
+    (
+        ([1234, b'abcd'], {}),
+        ([1234], {'b': b'abcd'}),
+        ([], {'a': 1234, 'b': b'abcd'}),
+        ([], {'b': b'abcd', 'a': 1234}),
+    ),
+)
+def test_contract_constructor_transact_with_constructor_with_arguments(
+        web3,
+        WithConstructorArgumentsContract,
+        WITH_CONSTRUCTOR_ARGUMENTS_RUNTIME,
+        constructor_args,
+        constructor_kwargs):
+    deploy_txn = WithConstructorArgumentsContract.constructor(
+        *constructor_args, **constructor_kwargs).transact()
+
+    txn_receipt = web3.eth.getTransactionReceipt(deploy_txn)
+    assert txn_receipt is not None
+
+    assert txn_receipt['contractAddress']
+    contract_address = txn_receipt['contractAddress']
+
+    blockchain_code = web3.eth.getCode(contract_address)
+    assert blockchain_code == decode_hex(WITH_CONSTRUCTOR_ARGUMENTS_RUNTIME)
+    assert EXPECTED_DATA_A == WithConstructorArgumentsContract(
+        address=contract_address).functions.data_a().call()
+    assert EXPECTED_DATA_B == WithConstructorArgumentsContract(
+        address=contract_address).functions.data_b().call()
```

Comment:
The global expected value is problematic, because it means the parametrization cannot test a variety of A and B values. It's better to put the expected values as their own parametrized values.